hash/hash_blake2b.c

```diff
@@ -0,0 +1,378 @@
+/**
+ * @addtogroup hash Hash algorithms
+ * @{
+ * @file
+ * @author [Daniel Murrow](https://github.com/dsmurrow)
+ * @brief [Blake2b cryptographic hash
+ * function](https://www.rfc-editor.org/rfc/rfc7693)
+ *
+ * The Blake2b cryptographic hash function provides 
+ * hashes for data that are secure enough to be used in 
+ * cryptographic applications. It is designed to perform 
+ * optimally on 64-bit platforms. The algorithm can output
+ * digests between 1 and 64 bytes long, for messages up to
+ * 128 bits in length. Keyed hashing is also supported for
+ * keys up to 64 bytes in length.
+ */
+#include <assert.h> /// for asserts
+#include <inttypes.h> /// for fixed-width integer types e.g. uint64_t and uint8_t
+#include <stdio.h> /// for IO
+#include <stdlib.h> /// for malloc, calloc, and free. As well as size_t
+
+/* Warning suppressed is in blake2b() function, more
+ * details are over there */
+#ifdef __GNUC__
+#pragma GCC diagnostic ignored "-Wshift-count-overflow"
+#elif _MSC_VER
+#pragma warning(disable : 4293)
+#endif
+
+/**
+ * @define bb
+ * @brief the size of a data block in bytes
+ */
+#define bb 128
+
+/**
+ * @define CEIL
+ * @brief ceiling division macro without floats
+ *
+ * @param a dividend
+ * @param divisor
+ */
+#define CEIL(a, b) (((a) / (b)) + ((a) % (b) != 0))
+
+/**
+ * @define MIN
+ * @brief returns minimum value
+ */
+#define MIN(a, b) ((a) < (b) ? (a) : (b))
+
+/**
+ * @define MAX
+ * @brief returns maximum value
+ */
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+
+/**
+ * @define ROTR64
+ * @brief macro to rotate 64-bit ints to the right
+ * Ripped from RFC 7693
+ */
+#define ROTR64(n, offset) (((n) >> (offset)) ^ ((n) << (64 - (offset))))
+
+/**
+ * @var R1, R2, R3, R4
+ * @brief rotation constants for mixing function G
+ */
+static const uint8_t R1 = 32, R2 = 24, R3 = 16, R4 = 63;
```

Comment:
AFAIK, Doxygen won't detect this properly in the documentation (see this [StackOverflow question](https://stackoverflow.com/a/15147744/14539444)).
The variables should be documented as shown in the answer or this (preferred way).

```c
uint32_t const my_variable = 132;      ///< Variable description
int16_t second_variable = 1;          ///< Variable description
```